Add support for `__CPROVER_DYNAMIC_OBJECT` to incremental SMT decision procedure

thomasspriggs opened the pull request:
This PR adds support for `__CPROVER_DYNAMIC_OBJECT` to incremental SMT decision procedure. Note that it is currently based on top of https://github.com/diffblue/cbmc/pull/7311 . So the first commit titled "Add unit test of object size tracking" should be reviewed as part of that separate PR, rather than this one.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add is_dynamic to object map

So that the information is available in subsequent conversions.

thomasspriggs committed:
Fix un-initialised fields in decision_procedure_objectt

The resolves a potential issue which clang-tidy warns about.

thomasspriggs committed:
Add `smt_is_dynamic_objectt` function

The SMT function which this commit implements declaration / application
/ definition will be used in follow up commits in order to convert
dynamic_object expressions to smt terms.

thomasspriggs committed:
Send is_dynamic object declaration and defintion to solver

In preparation for calling the new function in the conversion of
is_dynamic_object expressions to SMT terms.

thomasspriggs committed:
Add conversion of `is_dynamic_object_exprt` to SMT

This is the final step required to complete the implementation of the
support for this kind of expression.

thomasspriggs committed:
Add regression test of __CPROVER_DYNAMIC_OBJECT

TGWDB commented on src/solvers/smt2_incremental/object_tracking.cpp:
```diff
@@ -77,6 +79,21 @@ smt_object_mapt initial_smt_object_map()
+  // This check corresponds to the symbols created in
+  // `goto_symext::symex_allocate`, which implements the `__CPROVER_allocate`
+  // intrinsic function used by the standard library models.
+  bool dynamic_type = object.type().get_bool(ID_C_dynamic);
+  const auto symbol = expr_try_dynamic_cast<symbol_exprt>(object);
+  bool symbol_is_dynamic =
+    symbol &&
+    has_prefix(id2string(symbol->get_identifier()), SYMEX_DYNAMIC_PREFIX);
+  return dynamic_type || symbol_is_dynamic;
```
This seems like it is overly inefficient, can't we return on the `dynamic_type` immediately and then only progress to the casting and string prefix check if necessary?

thomasspriggs replied:
Yes, we could. I was thinking about readability rather than optimisation when I wrote this.

:thinking: `has_prefix`, `id2string` and `get_identifier` are all header defined functions. An optimising compiler may potentially be able to work out that these functions are all side-effect free and add in the early return...

thomasspriggs replied:
I've added the suggested optimisation for the moment.

esteffin commented on src/solvers/smt2_incremental/smt_is_dynamic_object.h:
```diff
@@ -0,0 +1,26 @@
+    smt_function_application_termt::factoryt<smt_command_functiont>;
+  make_applicationt make_application;
+  /// Makes the command to define the resulting \p size of calling the is
+  /// dynamic object function with \p unique_id.
```
⛏️  Change `is dynamic object` to `is_dynamic_object` as that's the function name

esteffin commented on src/solvers/smt2_incremental/smt_is_dynamic_object.h:
```diff
@@ -0,0 +1,26 @@
+#include <solvers/smt2_incremental/ast/smt_commands.h>
+#include <solvers/smt2_incremental/ast/smt_terms.h>
+
+struct smt_is_dynamic_objectt final
```
💡 Documentation inside looks very good, why not dropping a line to document the class itself?

thomasspriggs replied:
Added.

esteffin commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.h:
```diff
@@ -159,6 +160,7 @@ class smt2_incremental_decision_proceduret final
   /// stateful because it depends on the configuration of the number of object
   /// bits and how many bits wide the size type is configured to be.
   smt_object_sizet object_size_function;
+  smt_is_dynamic_objectt is_dynamic_object_function;
```
⛏️ Add documentation

thomasspriggs replied:
Added.

esteffin commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -255,6 +255,7 @@ smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
     smt_set_option_commandt{smt_option_produce_modelst{true}});
   solver_process->send(smt_set_logic_commandt{smt_logic_allt{}});
   solver_process->send(object_size_function.declaration);
+  solver_process->send(is_dynamic_object_function.declaration);
```
⛏️ Personal opinion: I know `smt_is_dynamic_objectt` is default-constructed, but readability would benefit if we explicitly construct it 5 lines above as in `object_map{initial_smt_object_map()}`.

esteffin commented on unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -464,6 +466,10 @@ TEST_CASE(
   const auto invalid_pointer_object_size_definition =
     test.object_size_function.make_definition(
       1, smt_bit_vector_constant_termt{0, 32});
+  const auto null_object_dynamic_definition =
+    test.is_dynamic_object_function.make_definition(0, false);
```
❓ Where do we handle the `null_object_dynamic_definition` case?

thomasspriggs replied:
It is currently defaulted to false in the definition of `decision_procedure_objectt`. I should set it explicitly in the `make_null_object` function.

thomasspriggs replied:
Done.

esteffin commented on regression/cbmc-incr-smt2/dynamic-memory/assert_dynamic.c:
```diff
@@ -0,0 +1,17 @@
+#include <assert.h>
```
⛏️ As we have a test checking that the object is not dynamic, should we avoid the nondeterministic boolean so that one verification leads to `FAILURE` and one to `SUCCESS`?
Otherwise we never check for `SUCCESS` (`UNSAT` on the SMT level).

esteffin reviewed (approved):
Looks good to me (with few minor suggestions)

thomasspriggs committed:
Rename `define_object_sizes` to `define_object_properties`

Because this now covers both the size and the definition of whether the
object is dynamic or not.
